airflow/bin/cli.py

```diff
@@ -233,35 +234,96 @@ def run(args):
-
-            # remove old logs (since they are already in S3)
-            if old_log:
-                new_log.replace(old_log, '')
-
-            try:
-                s3 = boto.connect_s3()
-                s3_key = boto.s3.key.Key(s3.get_bucket(bucket), key)
-
-                # append new logs to old S3 logs, if available
-                if s3_key.exists():
-                    old_s3_log = s3_key.get_contents_as_string().decode()
-                    new_log = old_s3_log + '\n' + new_log
-
-                # send log to S3
-                encrypt = configuration.get('core', 'ENCRYPT_S3_LOGS')
-                s3_key.set_contents_from_string(new_log, encrypt_key=encrypt)
-            except:
-                print('Could not send logs to S3.')
-
+    # check for deprecated configuration
+    remote_log_base = configuration.get('core', 'REMOTE_BASE_LOG_FOLDER')
+    if not remote_log_base and configuration.get('core', 'S3_LOG_FOLDER'):
+        warnings.warn(
+            'The S3_LOG_FOLDER configuration key has been replaced by '
+            'REMOTE_BASE_LOG_FOLDER. Your configuration still works but please '
+            'update airflow.cfg to ensure future compatibility.',
+            DeprecationWarning)
+        remote_log_base = configuration.get('core', 'S3_LOG_FOLDER')
+
+    if remote_log_base.startswith('s3:/'):
+        store_s3_logs(log_filename=filename, old_log=old_log)
+    elif remote_log_base.startswith('gs:/'):
+        store_gcs_logs(log_filename=filename, old_log=old_log)
+    elif remote_log_base:
+        raise ValueError(
+            'Unsupported remote log location: {}'.format(remote_log_base))
+
+def store_s3_logs(log_filename, old_log):
+    """
+    Store logs in S3
+    """
+    import boto
+    remote_log_base = configuration.get('core', 'REMOTE_BASE_LOG_FOLDER')
+    log_location = os.path.expanduser(
+        configuration.get('core', 'BASE_LOG_FOLDER'))
+    remote_log = log_filename.replace(log_location, remote_log_base)
+    bucket, key = remote_log.lstrip('s3:/').split('/', 1)
+    if os.path.exists(log_filename):
+
+        # get logs
+        with open(log_filename, 'r') as logfile:
+            new_log = logfile.read()
+
+        # remove old logs (since they are already in the remote location)
+        if old_log:
+            new_log.replace(old_log, '')
+
+        try:
+            s3 = boto.connect_s3()
+            s3_key = boto.s3.key.Key(s3.get_bucket(bucket), key)
+
+            # append new logs to old S3 logs, if available
+            if s3_key.exists():
+                old_s3_log = s3_key.get_contents_as_string().decode()
+                new_log = old_s3_log + '\n' + new_log
+
+            # send log to S3
+            encrypt = configuration.get('core', 'ENCRYPT_S3_LOGS')
+            s3_key.set_contents_from_string(new_log, encrypt_key=encrypt)
+        except:
+            print('Could not send logs to S3.')
+
+def store_gcs_logs(log_filename, old_log):
```

Comment:
The original intent (with S3) was to allow as minimal configuration as possible so that log backups "just worked" -- as long as AWS credentials were available on the host, logs would be shipped. All users had to do was put the location in `airflow.cfg`. With hooks, it requires not just modifying `airflow.cfg` but also creating connections, etc. Today it probably seems obvious to use a connection but back when Airflow was new it wasn't :)

So with that said, I'm not opposed to using hooks, but I prefer this implementation to the current `GCSHook`. Here we leverage the `gcloud` python library to move the logs in just a few lines of code; I think the `GCSHook` rolls its own REST API access and seems quite fragile (why not let Google maintain the API?). Internally, we have a `GCSHook` that is much closer to the `S3Hook` implementation, just wrapping `gcloud` instead of `boto`. This just exposes a little of that functionality.

However, this is compatible with service accounts, just define `GOOGLE_APPLICATION_CREDENTIALS=/path/to/service/key.json` in your environment.